subsys/net/lib/http/http_app_client.c

```diff
@@ -0,0 +1,720 @@
+/*
+ * Copyright (c) 2017 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#if defined(CONFIG_NET_DEBUG_HTTP)
+#if defined(CONFIG_HTTPS)
+#define SYS_LOG_DOMAIN "https/client"
+#else
+#define SYS_LOG_DOMAIN "http/client"
+#endif
+#define NET_SYS_LOG_LEVEL SYS_LOG_LEVEL_DEBUG
+#define NET_LOG_ENABLED 1
+#endif
+
+#include <zephyr.h>
+#include <string.h>
+#include <strings.h>
+#include <errno.h>
+#include <stdlib.h>
+#include <version.h>
+
+#include <net/net_core.h>
+#include <net/net_ip.h>
+#include <net/http.h>
+
+#include "../../ip/net_private.h"
+
+#define BUF_ALLOC_TIMEOUT 100
+
+#define RC_STR(rc)	(rc == 0 ? "OK" : "ERROR")
+
+#define HTTP_EOF           "\r\n\r\n"
+
+#define HTTP_HOST          "Host"
+#define HTTP_CONTENT_TYPE  "Content-Type"
+#define HTTP_CONTENT_LEN   "Content-Length"
+#define HTTP_CONT_LEN_SIZE 6
+
+int client_reset(struct http_ctx *ctx)
+{
+	http_parser_init(&ctx->http.parser, HTTP_RESPONSE);
+
+	memset(ctx->http.rsp.http_status, 0,
+	       sizeof(ctx->http.rsp.http_status));
+
+	ctx->http.rsp.cl_present = 0;
+	ctx->http.rsp.content_length = 0;
+	ctx->http.rsp.processed = 0;
+	ctx->http.rsp.body_found = 0;
+	ctx->http.rsp.message_complete = 0;
+	ctx->http.rsp.body_start = NULL;
+
+	memset(ctx->http.rsp.response_buf, 0, ctx->http.rsp.response_buf_len);
+	ctx->http.rsp.data_len = 0;
+
+	return 0;
+}
+
+int http_request(struct http_ctx *ctx, struct http_request *req, s32_t timeout)
+{
+	const char *method = http_method_str(req->method);
+	int ret;
+
+	if (ctx->pending) {
+		net_pkt_unref(ctx->pending);
+		ctx->pending = NULL;
+	}
+
+	http_add_header(ctx, method, ctx->http.req.user_data);
+	http_add_header(ctx, " ", ctx->http.req.user_data);
+	http_add_header(ctx, req->url, ctx->http.req.user_data);
+	http_add_header(ctx, req->protocol, ctx->http.req.user_data);
+	http_add_header(ctx, HTTP_CRLF, ctx->http.req.user_data);
+
+	if (req->host) {
+		ret = http_add_header_field(ctx, HTTP_HOST, req->host,
+					    ctx->http.req.user_data);
+		if (ret < 0) {
+			goto out;
+		}
+	}
+
+	if (req->header_fields) {
+		ret = http_add_header(ctx, req->header_fields,
+				      ctx->http.req.user_data);
+		if (ret < 0) {
+			goto out;
+		}
+	}
+
+	if (req->content_type_value) {
+		ret = http_add_header_field(ctx, HTTP_CONTENT_TYPE,
+					    req->content_type_value,
+					    ctx->http.req.user_data);
+		if (ret < 0) {
+			goto out;
+		}
+	}
+
+	if (req->payload && req->payload_size) {
+		char content_len_str[HTTP_CONT_LEN_SIZE];
+		int i;
+
+		ret = snprintk(content_len_str, HTTP_CONT_LEN_SIZE,
+			       "%u", req->payload_size);
+		if (ret <= 0 || ret >= HTTP_CONT_LEN_SIZE) {
+			ret = -ENOMEM;
+			goto out;
+		}
+
+		ret = http_add_header_field(ctx, HTTP_CONTENT_LEN,
+					    content_len_str,
+					    ctx->http.req.user_data);
+		if (ret < 0) {
+			goto out;
+		}
+
+		ret = http_add_header(ctx, HTTP_CRLF, ctx->http.req.user_data);
+		if (ret < 0) {
+			goto out;
+		}
+
+		for (i = 0; i < req->payload_size;) {
```

Comment:
This loop should not be needed (chunked encoding itself doesn't have limitations on chunk size, right?). http_send_chunk() uses http_prepare_and_send(), and that guarantees that it sends all the data (unless error occurs). That's the beauty of it (the downside is the need to check each call for error - but that's more the errors to blame, not the approach ;-).

Reply:
But it's not needed ;-). Let's spend those code bytes on adding error checking consistently everywhere.

Reply:
The API has now http_send_chunk() which adds a chunk. If we want to have only one chunk in payload regardless of size, then this API is not useful and the caller would need to know the total size of the sent data in advance. So I propose that we go with this way (multiple chunks) atm.

Reply:
Well, we actually should rework HTTP request API to allow to do a streaming POST for example ;-). I don't raise that now, because there're enough improvements for this patch already. OTOH, think about that now will allow to deprecate less API calls later. So, let me know if you'd like to discuss that or will have thought about that yourself.


Otherwise, we know that http_send_chunk(..., len) will return len (that's the API contract), so the loop will run only once. We can leave it until next time of course.